base/iobuffer.jl

```diff
@@ -1,26 +1,117 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-## work with AbstractVector{UInt8} via I/O primitives ##
+# IOBuffer is a Memory{UInt8} backed IO type for in-memory IO.
+
+# Here, u represents used bytes (already read), X represents bytes still to read,
+# - represents bytes uninitialized data but which can be written to later.
+# . represents bytes before offset, which the buffer will not touch, until
+# a write operation happens.
+
+#   .....uuuuuuuuuuuuuXXXXXXXXXXXXX------------
+#   |   |            |           |            |    |
+#   |   offset       ptr         size         |    maxsize (≥ lastindex)
+#   1                                         lastindex(data)
+
+# N.B: `mark` does not correspond to any index in the buffer, but is instead equal
+# to position (== io.offset + io.ptr - 1) at the time it is set.
+
+#            AFTER COMPACTION
+
+#   uuuuuXXXXXXXXXXXXX---------------------
+#  ||    |           |                    |    |
+#  |1    ptr         size                 |    maxsize (≥ lastindex)
+#  |                                      lastindex(data)
+#  offset (set to zero)
+
+# * The underlying array is always 1-indexed
+# * The IOBuffer has full control (ownership) of the underlying array, only when
+#   buffer.write == true.
+# * Data before the mark can be deleted, shifting the whole thing to the left
+#   to make room for more data, without replacing or resizing data.
+#   This can be done only if the buffer is not seekable
 
-# Stateful string
 mutable struct GenericIOBuffer{T<:AbstractVector{UInt8}} <: IO
-    data::T # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
-    reinit::Bool # if true, data needs to be re-allocated (after take!)
+    # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
```

Comment:
```suggestion
    # T should support: getindex, setindex!, length, copyto!, similar, size, and (optionally) resize!
```

I noticed you added `lastindex` to the list of functions used. That is just a fairly convoluted way of calling `length` though, so I don't really know if that is necessary, but the `require_one_based_indexing` someone previously added already does the same dispatch calls, so it seems okay to just add that to the list here.